### PR TITLE
Validate zero vector when using cosine metric

### LINKED ISSUE #1461: Validate the vector when using cosine metric
**Is your feature request related to a problem?**
When I use cosine metric and provide zero vector(all elements are zero) as query vector, I get no hits, then I realize that a zero vector is meaningless for cosine distance, but OpenSearch dose not provide any information about this, which may confuse users.

**What solution would you like?**
Validate the vector passed by the user during querying or indexing, if the vector is zero vector and the space type of index is cosine, an exception will be thrown. If you guys believe it is necessary, I can create a PR. Any thoughts? 

**What alternatives have you considered?**

**Do you have any additional context?**

### PR DESCRIPTION
**Pull request by bugmakerrrrrr**
### Description
Validate the vector passed by the user during querying or indexing, if the vector is zero vector and the space type of index is cosine, an exception will be thrown.

**Examples**
1. When we use cosine space type and index a zero vector through bulk api, the response returned will be as follows:
```
{
    "took": 2,
    "errors": true,
    "items": [
        {
            "index": {
                "_index": "products-shirts",
                "_id": "1",
                "status": 400,
                "error": {
                    "type": "mapper_parsing_exception",
                    "reason": "failed to parse field [item_vector] of type [knn_vector] in document with id '1'. Preview of field's value: 'null'",
                    "caused_by": {
                        "type": "illegal_argument_exception",
                        "reason": "zero vector is not supported when space type is [cosinesimil]"
                    }
                }
            }
        }
    ]
}
```
2. If we use cosine metric and query index is a zero vector, the response returned will be as following:
```
{
    "error": {
        "root_cause": [
            {
                "type": "query_shard_exception",
                "reason": "failed to create query: zero vector is not supported when space type is [cosinesimil]",
                "index": "products-shirts",
                "index_uuid": "5ccs3V8rRceXFOaMAvDtMw"
            }
        ],
        "type": "search_phase_execution_exception",
        "reason": "all shards failed",
        "phase": "query",
        "grouped": true,
        "failed_shards": [
            {
                "shard": 0,
                "index": "products-shirts",
                "node": "Fq1_A7lHSAu25Beapfb5-A",
                "reason": {
                    "type": "query_shard_exception",
                    "reason": "failed to create query: zero vector is not supported when space type is [cosinesimil]",
                    "index": "products-shirts",
                    "index_uuid": "5ccs3V8rRceXFOaMAvDtMw",
                    "caused_by": {
                        "type": "illegal_argument_exception",
                        "reason": "zero vector is not supported when space type is [cosinesimil]"
                    }
                }
            }
        ]
    },
    "status": 400
}
```

3. When we use euclidean space type and index a zero vector through bulk api, the response returned will be as follows:
```
{
    "took": 301,
    "errors": false,
    "items": [
        {
            "index": {
                "_index": "products-shirts",
                "_id": "12",
                "_version": 1,
                "result": "created",
                "_shards": {
                    "total": 1,
                    "successful": 1,
                    "failed": 0
                },
                "_seq_no": 11,
                "_primary_term": 1,
                "status": 201
            }
        }
    ]
}
```

4. If we use euclidean metric and query index is a zero vector, the response returned will be as following:
```
{
    "took": 2,
    "timed_out": false,
    "_shards": {
        "total": 1,
        "successful": 1,
        "skipped": 0,
        "failed": 0
    },
    "hits": {
        "total": {
            "value": 2,
            "relation": "eq"
        },
        "max_score": 1.0,
        "hits": [
            {
                "_index": "products-shirts",
                "_id": "12",
                "_score": 1.0,
                "_source": {
                    "item_vector": [
                        0,
                        0,
                        0
                    ],
                    "size": "large",
                    "rating": 3
                }
            },
            {
                "_index": "products-shirts",
                "_id": "11",
                "_score": 0.055555556,
                "_source": {
                    "item_vector": [
                        3,
                        2,
                        2
                    ],
                    "size": "large",
                    "rating": 6
                }
            }
        ]
    }
}
```
 
### Issues Resolved
#1461
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed as per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by jmazanec15** [comment]
Thanks @bugmakerrrrrr could you update the Changelog as well? https://github.com/opensearch-project/k-NN/blob/main/CHANGELOG.md

File: src/main/java/org/opensearch/knn/common/KNNVectorUtil.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+    /**
+     * Check if all the elements of a given vector are zero
+     *
+     * @param vector the vector
+     * @return true if yes; otherwise false
+     */
+    public static boolean isZeroVector(byte[] vector) {
+        for (byte e : vector) {
+            if (e != (byte) 0) {
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/common/KNNVectorUtil.java:19]: qq: Do we need 0 cast?

**Review by jmazanec15 — COMMENTED** [review]
Overall looks good to me. Thanks for this! Will wait to approve on changelog verification

File: src/main/java/org/opensearch/knn/common/KNNVectorUtil.java (hunk 1 above)
Comment (bugmakerrrrrr) [reply src/main/java/org/opensearch/knn/common/KNNVectorUtil.java:19]: not necessary, it's just my habit. I can remove it.

**Review by jmazanec15 — APPROVED** [review]
LGTM - thanks for the change!

**Comment by navneet1v** [comment]
@bugmakerrrrrr please add the successful indexing and query I/O either in description or reply to this message to provide information what is the experience now.

File: src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapperUtil.java [hunk 2]
```diff
@@ -74,6 +76,34 @@ public static void validateByteVectorValue(float value) {
...
+        }
+    }
+
+    /**
+     * Validate if the given float vector is supported by the given space type
+     *
+     * @param vector     the given vector
+     * @param spaceType  the given space type
+     */
+    public static void validateFloatVector(float[] vector, SpaceType spaceType) {
```
Comment (navneet1v) [inline src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapperUtil.java:99]: should we make these parts of SpaceType enum class? Also, does these validations true for all engines?

File: src/main/java/org/opensearch/knn/common/KNNVectorUtil.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.common;
+
+public class KNNVectorUtil {
+    private KNNVectorUtil() {}
```
Comment (navneet1v) [inline src/main/java/org/opensearch/knn/common/KNNVectorUtil.java:9]: use @NoArgsConstructor(access = Access.PRIVATE)

File: src/main/java/org/opensearch/knn/common/KNNVectorUtil.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        return true;
+    }
+
+    /**
+     * Check if all the elements of a given vector are zero
+     *
+     * @param vector the vector
+     * @return true if yes; otherwise false
+     */
+    public static boolean isZeroVector(float[] vector) {
```
Comment (navneet1v) [inline src/main/java/org/opensearch/knn/common/KNNVectorUtil.java:32]: lets validate that these arrays are no null

**Review by martin-gaievski — APPROVED** [review]
Looks good to me, thank you

File: src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapper.java [hunk 5]
```diff
@@ -404,22 +421,24 @@ public KNNVectorFieldType(
...
             String name,
             Map<String, String> meta,
             int dimension,
-            KNNMethodContext knnMethodContext,
-            String modelId,
-            VectorDataType vectorDataType
+            @Nullable KNNMethodContext knnMethodContext,
+            @Nullable String modelId,
+            VectorDataType vectorDataType,
+            @Nullable SpaceType spaceType
```
Comment (navneet1v) [inline src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapper.java:434]: why we are making these args as Nullables?

File: src/main/java/org/opensearch/knn/index/query/KNNQueryBuilder.java [hunk 6]
```diff
@@ -284,12 +287,15 @@ protected Query doToQuery(QueryShardContext context) {
         KNNMethodContext knnMethodContext = knnVectorFieldType.getKnnMethodContext();
         KNNEngine knnEngine = KNNEngine.DEFAULT;
         VectorDataType vectorDataType = knnVectorFieldType.getVectorDataType();
+        SpaceType spaceType = knnVectorFieldType.getSpaceType();
 
         if (fieldDimension == -1) {
             // If dimension is not set, the field uses a model and the information needs to be retrieved from there
+            assert spaceType == null;
```
Comment (navneet1v) [inline src/main/java/org/opensearch/knn/index/query/KNNQueryBuilder.java:294]: 1. Putting assert here is dangerous as this will lead to exception messages not understood by users. Lets make sure that we are making sure proper exceptions are thrown here.

2. Due to this change the comment location have been moved. Lets ensure the comments is at right position.

File: src/main/java/org/opensearch/knn/index/query/KNNQueryBuilder.java [hunk 7]
```diff
@@ -308,6 +314,9 @@ protected Query doToQuery(QueryShardContext context) {
                 validateByteVectorValue(vector[i]);
                 byteVector[i] = (byte) vector[i];
             }
+            validateByteVector(byteVector, spaceType);
```
Comment (navneet1v) [inline src/main/java/org/opensearch/knn/index/query/KNNQueryBuilder.java:317]: Can we do this validation while reading the vectors in above loop?

File: src/test/java/org/opensearch/knn/index/OpenSearchIT.java [hunk 8]
```diff
@@ -263,6 +265,35 @@ public void testIndexingVectorValidation_differentSizes() throws Exception {
         assertThat(EntityUtils.toString(ex.getResponse().getEntity()), containsString("Vector dimension mismatch. Expected: 4, Given: 5"));
     }
 
+    @SneakyThrows
+    public void testIndexingVectorValidation_zeroVector() {
```
Comment (navneet1v) [inline src/test/java/org/opensearch/knn/index/OpenSearchIT.java:269]: This test should be part of Either Lucene IT or IndexingIT. This class is a class for containing common functions. Lets not add tests where which runs.

**Comment by navneet1v** [comment]
@bugmakerrrrrr can we add a BWC test where we try to add 0 vectors in older version and then zero vectors in upgraded version.

Same for the query too.

File: src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapper.java [hunk 9]
```diff
@@ -281,7 +285,7 @@ public KNNVectorFieldMapper build(BuilderContext context) {
 
                 return new ModelFieldMapper(
                     name,
-                    new KNNVectorFieldType(buildFullName(context), metaValue, -1, knnMethodContext, modelIdAsString),
+                    new KNNVectorFieldType(buildFullName(context), metaValue, -1, modelIdAsString),
```
[the remark this answers is not in the document]
Comment (bugmakerrrrrr) [reply src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapper.java:288]: `knnMethodContext` is always null for `ModelFieldMapper`, I believe that this change will make it more clear

File: src/main/java/org/opensearch/knn/index/query/KNNQueryBuilder.java [hunk 10]
```diff
@@ -34,7 +35,9 @@
 import java.util.Objects;
 
 import static org.opensearch.knn.index.IndexUtil.isClusterOnOrAfterMinRequiredVersion;
+import static org.opensearch.knn.index.mapper.KNNVectorFieldMapperUtil.validateByteVector;
 import static org.opensearch.knn.index.mapper.KNNVectorFieldMapperUtil.validateByteVectorValue;
+import static org.opensearch.knn.index.mapper.KNNVectorFieldMapperUtil.validateFloatVector;
```
[the remark this answers is not in the document]
Comment (bugmakerrrrrr) [reply src/main/java/org/opensearch/knn/index/query/KNNQueryBuilder.java:40]: make sense. I can pull these out into a separate `KNNValidationUtil` class

File: src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapperUtil.java (hunk 2 above)
Comment (bugmakerrrrrr) [reply src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapperUtil.java:99]: the zero vector validation is same for all engines, but maybe not true for other validations introduced in the future. For example, the `dot product` metric in Lucene requires all vectors must be normalized, but Faiss doesn't, so I tend to keep these part in this class, any thoughts?

File: src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapper.java (hunk 5 above)
Comment (bugmakerrrrrr) [reply src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapper.java:434]: these parameters can be null for different mapper, this annotation can make it more clear.

File: src/main/java/org/opensearch/knn/index/query/KNNQueryBuilder.java (hunk 7 above)
Comment (bugmakerrrrrr) [reply src/main/java/org/opensearch/knn/index/query/KNNQueryBuilder.java:317]: we cannot reuse the validation function or short circuit like `VectorUtil.isZeroVector` if we implement it in the loop. I prefer not to do so

File: src/test/java/org/opensearch/knn/index/OpenSearchIT.java (hunk 8 above)
Comment (bugmakerrrrrr) [reply src/test/java/org/opensearch/knn/index/OpenSearchIT.java:269]: I think this test is similar to `OpenSearchIT#testIndexingVectorValidation_differentSizes`, and it is not limited to the Lucene engine, but can also be applied to other engines.

File: src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapper.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (navneet1v) [reply src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapper.java:288]: I would recommend keeping it as is. This will ensure that changes are minimal for this PR.

File: src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapperUtil.java (hunk 2 above)
Comment (navneet1v) [reply src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapperUtil.java]: if its true for all engines that it makes more sense to add as part of enum class.

**Comment by navneet1v** [comment]
> > @bugmakerrrrrr can we add a BWC test where we try to add 0 vectors in older version and then zero vectors in upgraded version.
> > Same for the query too.
> 
> why do we need a BWC test? I think this a breaking change

if this is a breaking change, then what are the steps that users needs to take so that they can safely upgrade?

cc: @jmazanec15

**Comment by bugmakerrrrrr** [comment]
> if this is a breaking change, then what are the steps that users needs to take so that they can safely upgrade?
> 
If the user upgrades the cluster, any request containing a zero vector will be rejected and fail when using cosine metric. And during the upgrade process, requests containing zero vectors will succeed on the old nodes but will be rejected on the upgraded nodes. 

If the user uses non-zero vectors under cosine metric, there will be no impact. However, if the user uses zero vectors, it may be considered as an abuse or misconfiguration, and request failure may be acceptable.

**Comment by jmazanec15** [comment]
I agree with @bugmakerrrrrr . From https://github.com/opensearch-project/k-NN/issues/1461#issuecomment-1929768908, the error is:
```
org.opensearch.action.search.SearchPhaseExecutionException: all shards failed
	at org.opensearch.action.search.AbstractSearchAsyncAction.onPhaseFailure(AbstractSearchAsyncAction.java:665) [opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:373) [opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.action.search.AbstractSearchAsyncAction.onPhaseDone(AbstractSearchAsyncAction.java:704) [opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.action.search.AbstractSearchAsyncAction.onShardFailure(AbstractSearchAsyncAction.java:473) [opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.action.search.AbstractSearchAsyncAction$1.onFailure(AbstractSearchAsyncAction.java:295) [opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.action.search.SearchExecutionStatsCollector.onFailure(SearchExecutionStatsCollector.java:104) [opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.action.ActionListenerResponseHandler.handleException(ActionListenerResponseHandler.java:74) [opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.action.search.SearchTransportService$ConnectionCountingHandler.handleException(SearchTransportService.java:755) [opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.transport.TransportService$6.handleException(TransportService.java:884) [opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:1504) [opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.transport.TransportService$DirectResponseChannel.processException(TransportService.java:1618) [opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.transport.TransportService$DirectResponseChannel.sendResponse(TransportService.java:1592) [opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.transport.TaskTransportChannel.sendResponse(TaskTransportChannel.java:79) [opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.transport.TransportChannel.sendErrorResponse(TransportChannel.java:71) [opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.action.support.ChannelActionListener.onFailure(ChannelActionListener.java:70) [opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.action.ActionRunnable.onFailure(ActionRunnable.java:103) [opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:54) [opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.threadpool.TaskAwareRunnable.doRun(TaskAwareRunnable.java:78) [opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52) [opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:102) [opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:908) [opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52) [opensearch-2.9.0.jar:2.9.0]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: org.opensearch.OpenSearchException$3: Index 2147483647 out of bounds for length 12
	at org.opensearch.OpenSearchException.guessRootCauses(OpenSearchException.java:685) ~[opensearch-core-2.9.0.jar:2.9.0]
	at org.opensearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:371) [opensearch-2.9.0.jar:2.9.0]
	... 23 more
Caused by: java.lang.IndexOutOfBoundsException: Index 2147483647 out of bounds for length 12
	at jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64) ~[?:?]
	at jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70) ~[?:?]
	at jdk.internal.util.Preconditions.checkIndex(Preconditions.java:248) ~[?:?]
	at java.util.Objects.checkIndex(Objects.java:372) ~[?:?]
	at org.apache.lucene.index.SegmentReader.document(SegmentReader.java:252) ~[lucene-core-9.7.0.jar:9.7.0 ccf4b198ec328095d45d2746189dc8ca633e8bcf - 2023-06-21 11:48:16]
	at org.apache.lucene.index.FilterLeafReader.document(FilterLeafReader.java:404) ~[lucene-core-9.7.0.jar:9.7.0 ccf4b198ec328095d45d2746189dc8ca633e8bcf - 2023-06-21 11:48:16]
	at org.apache.lucene.index.FilterLeafReader.document(FilterLeafReader.java:404) ~[lucene-core-9.7.0.jar:9.7.0 ccf4b198ec328095d45d2746189dc8ca633e8bcf - 2023-06-21 11:48:16]
	at org.opensearch.search.fetch.FetchPhase.loadStoredFields(FetchPhase.java:563) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.search.fetch.FetchPhase.prepareNonNestedHitContext(FetchPhase.java:337) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.search.fetch.FetchPhase.prepareHitContext(FetchPhase.java:298) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.search.fetch.FetchPhase.execute(FetchPhase.java:169) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.search.SearchService.executeFetchPhase(SearchService.java:628) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.search.SearchService.executeQueryPhase(SearchService.java:603) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.search.SearchService$2.lambda$onResponse$0(SearchService.java:565) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.action.ActionRunnable.lambda$supply$0(ActionRunnable.java:73) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.action.ActionRunnable$2.doRun(ActionRunnable.java:88) ~[opensearch-2.9.0.jar:2.9.0]
	at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52) ~[opensearch-2.9.0.jar:2.9.0]
	... 8 more
```

The `Caused by: java.lang.IndexOutOfBoundsException: Index 2147483647 out of bounds for length 12` is concerning because it looks like it is returning garbage which can cause worse exceptions. Of more concern, this could cause a divide by 0 exception in native code, which would lead to node drop.

In terms of BWC, if a shard containing a 0 vector is moved to a new node, it wont fail to migrate - we dont check. It will only fail if a user attempts to reindex. That being said, I am not concerned about breaking BWC.

**Review by jmazanec15 — APPROVED** [review]

**Comment by navneet1v** [comment]
So what we are suggesting here is, if a user has add a zero vector already in the index after upgrade we are okay to break the customer search query as zero vector has potential to go with divide by 0 exceptions.

**Comment by navneet1v** [comment]
@bugmakerrrrrr Code looks good to me. Please resolve the conflicts in Changelog.md file.

**Comment by jmazanec15** [comment]
> So what we are suggesting here is, if a user has add a zero vector already in the index after upgrade we are okay to break the customer search query as zero vector has potential to go with divide by 0 exceptions.

We would fail the query only if query vector is 0 (not if index vectors are 0). If it is 0, it will have meaningless results anyway.

**Comment by jmazanec15** [comment]
@navneet1v added doc issue: https://github.com/opensearch-project/documentation-website/issues/6678.

**Review by jmazanec15 — APPROVED** [review]

**Review by navneet1v — APPROVED** [review]
